fix(core): use appropriate inert document strategy for Firefox & Safari

petebacondarwin opened the pull request:
Both Firefox and Safari are vulnerable to XSS if we use an inert document
created via `document.implementation.createHTMLDocument()`.

Now we check for those vulnerabilities and then use a DOMParser or XHR
strategy if needed.

Thanks to @cure53 for the heads up on this issue.

petebacondarwin commented:
Locally I am getting this warning:

```
LOG: 'Attempting to configure '__esModule' with descriptor '{"value":true}' on object 'undefined' and got error, giving up: TypeError: Object.defineProperty called on non-object'
```

I am not sure what is causing this.

mary-poppins commented:
The angular.io preview for abc704a9b03074406c93307a10e6e31cea6b36dd is available [here][1].

[1]: https://pr17019-abc704a9b03074406c93307a10e6e31cea6b36dd.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 5c0104d99376ce404211333ff200937c4b7bd50d is available [here][1].

[1]: https://pr17019-5c0104d99376ce404211333ff200937c4b7bd50d.ngbuilds.io/

mary-poppins commented:
The angular.io preview for bef4d3392b86c7291c78df5c66586781107a10a4 is available [here][1].

[1]: https://pr17019-bef4d3392b86c7291c78df5c66586781107a10a4.ngbuilds.io/

mhevery commented on packages/platform-browser/test/security/html_sanitizer_spec.ts:
```diff
@@ -134,6 +134,22 @@ export function main() {
+    it('should not allow JavaScript execution when creating inert document', () => {
+      sanitizeHtml(defaultDoc, '<svg><g onload="window.xxx = 100"></g></svg>');
+      expect((window as any).xxx).toBe(undefined);
+      delete (window as any).xxx;
```
Please don't do `delete` it will make window a slow object for the rest of the tests. Just assign `undefined`

petebacondarwin replied:
OK!

petebacondarwin replied:
Although that is quite an ironic statement given how slow the tests are generally :-P

petebacondarwin commented:
The unit tests are failing on Travis. It seems it is because the `inertBodyElement` does not have the `querySelector` method. Is this some different mode than I am running locally?

mary-poppins commented:
The angular.io preview for 3c1adac41b8a17262ae4c2dde7b61370869244f4 is available [here][1].

[1]: https://pr17019-3c1adac41b8a17262ae4c2dde7b61370869244f4.ngbuilds.io/

mprobst commented:
/CC @rjamet

mprobst commented on packages/platform-browser/src/security/html_sanitizer.ts:
```diff
@@ -121,53 +95,54 @@ class SanitizingHtmlSerializer {
   // because characters were re-encoded.
   public sanitizedSomething = false;
   private buf: string[] = [];
+  private DOM = getDOM();
```
nit: should be `lowerCamelCase`, i.e. `dom`, right? It doesn't have to be a field in the class instance btw, it should be an initialized-once module level field. We just need to take care to initialize after the initial angular load, otherwise we risk getting the wrong DOM.

petebacondarwin replied:
I was copying this casing from the previous code: https://github.com/angular/angular/pull/17019/files/3c1adac41b8a17262ae4c2dde7b61370869244f4#diff-4fba2f404f1048cbc43f747e3bc2606dL18. Happy to change this if appropriate.

I was not sure about the benefit of using a module level variable. Throughout the rest of the Angular code base, this `getDOM()` is used in a variety of ways: as it is needed (not cached at all) - e.g. the `Title` service; cached per instance - e.g. `Meta` service; and cached per module. But the majority use the call as required approach and looking at the implementation of this method it is super simple method:

```
export function getDOM() {
  return _DOM;
}
```

The downside of caching is that we get the wrong version of the DOM so it seems to me that it is actually safer and easier (and no slower) to just call `getDOM()` as required.

mprobst replied:
Indeed, we don't really have a system here. I think I'd either access it through the getter (which hopefully your optimizer should inline) or from a lazily initialized module variable. YMMV, and I don't have data pointing either way.

mprobst commented on packages/platform-browser/src/security/html_sanitizer.ts:
```diff
@@ -233,33 +208,20 @@ function encodeEntities(value: string) {
  * the DOM in a browser environment.
  */
 export function sanitizeHtml(defaultDoc: any, unsafeHtmlInput: string): string {
+  const DOM = getDOM();
```
initialize `DOM` as a module level field here, as you do with `inertBodyHelper` below

petebacondarwin replied:
The only reason I use a module level variable for `inertBodyHelper` is that it is used in a free standing exported function (rather than a class) and so there was no other container to hold the cached object.

mprobst commented on packages/platform-browser/src/security/inert_body.ts:
```diff
@@ -0,0 +1,127 @@
+
+import {DomAdapter, getDOM} from '../dom/dom_adapter';
+
+export class InertBodyHelper {
```
add a comment on what this does?

mprobst commented on packages/platform-browser/src/security/inert_body.ts:
```diff
@@ -0,0 +1,127 @@
+import {DomAdapter, getDOM} from '../dom/dom_adapter';
+
+export class InertBodyHelper {
+  private inertDocument = this.DOM.createHtmlDocument();
```
you should document the lifetime of this object, i.e. that we reuse the same document for many sanitization runs has performance implications (creating many documents would be expensive).

mprobst commented on packages/platform-browser/src/security/inert_body.ts:
```diff
@@ -0,0 +1,127 @@
+
+export class InertBodyHelper {
+  private inertDocument = this.DOM.createHtmlDocument();
+  private inertBodyElement = this.DOM.querySelector(this.inertDocument, 'body');
```
can't you access `inertDocument.body`?

mprobst commented on packages/platform-browser/src/security/inert_body.ts:
```diff
@@ -0,0 +1,127 @@
+    // Check for the Safari 10.1 bug - which allows JS to run inside the SVG G element
+    this.inertBodyElement.innerHTML = '<svg><g onload="this.parentNode.remove()"></g></svg>';
+    if (!this.inertBodyElement.querySelector('svg')) {
+      this.getInertBodyElement = this.getInertBodyElement_XHR;
```
return early here, then move the `else` block on level higher to reduce indent.

mprobst replied:
can you document that this is the Safari workaround case?

mprobst commented on packages/platform-browser/src/security/inert_body.ts:
```diff
@@ -0,0 +1,127 @@
+      this.inertBodyElement.innerHTML =
+          '<svg><p><style><img src="</style><img src=x onerror=alert(1)//">';
+      if (this.inertBodyElement.querySelector('svg img')) {
+        this.getInertBodyElement = this.getInertBodyElement_DOMParser;
```
can you document which of these cases is the workaround for the FF bug?

mprobst commented on packages/platform-browser/src/security/inert_body.ts:
```diff
@@ -0,0 +1,127 @@
+      if (this.inertBodyElement.querySelector('svg img')) {
+        this.getInertBodyElement = this.getInertBodyElement_DOMParser;
+      } else {
+        this.getInertBodyElement = this.getInertBodyElement_InertDocument;
```
document that this is the default and sane way.

mprobst commented on packages/platform-browser/src/security/inert_body.ts:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  /**
+   * Use createHtmlDocument to create and fill an inert body element
```
the code below doesn't use createHtmlDocument, fix the docs?

mprobst commented on packages/platform-browser/test/security/html_sanitizer_spec.ts:
```diff
@@ -134,6 +134,22 @@ export function main() {
 
+    // See
+    // https://github.com/cure53/DOMPurify/blob/a992d3a75031cb8bb032e5ea8399ba972bdf9a65/src/purify.js#L439-L449
+    it('should not allow JavaScript execution when creating inert document', () => {
```
For my information, are we actually running tests in Safari?

petebacondarwin replied:
I think we are:

https://github.com/angular/angular/blob/master/browser-providers.conf.js#L32
and
https://github.com/angular/angular/blob/master/browser-providers.conf.js#L53

koto commented:
LGTM security-wise.

petebacondarwin commented:
@mprobst I added another commit with changes based on your review.

mary-poppins commented:
The angular.io preview for 86914519a576ede177786f3051bc05e6bda62dba is available [here][1].

[1]: https://pr17019-86914519a576ede177786f3051bc05e6bda62dba.ngbuilds.io/

mary-poppins commented:
The angular.io preview for b0a1f13780440073b2a0bf50e17bc00650b544a1 is available [here][1].

[1]: https://pr17019-b0a1f13780440073b2a0bf50e17bc00650b544a1.ngbuilds.io/

mprobst reviewed (approved):
LGTM.

petebacondarwin commented:
Just got to work out why it doesn't look good to Travis :-(

mary-poppins commented:
You can preview da10f2f at https://pr17019-da10f2f.ngbuilds.io/.

mary-poppins commented:
You can preview 65ab3d5 at https://pr17019-65ab3d5.ngbuilds.io/.

mary-poppins commented:
You can preview 6723f8a at https://pr17019-6723f8a.ngbuilds.io/.

mary-poppins commented:
You can preview b17f6c7 at https://pr17019-b17f6c7.ngbuilds.io/.

mary-poppins commented:
You can preview cef5c98 at https://pr17019-cef5c98.ngbuilds.io/.

mary-poppins commented:
You can preview 21a3121 at https://pr17019-21a3121.ngbuilds.io/.

mary-poppins commented:
You can preview a2cac9e at https://pr17019-a2cac9e.ngbuilds.io/.

petebacondarwin committed:
fix(core): use appropriate inert document strategy for Firefox & Safari

Both Firefox and Safari are vulnerable to XSS if we use an inert document
created via `document.implementation.createHTMLDocument()`.

Now we check for those vulnerabilities and then use a DOMParser or XHR
strategy if needed.

Further the platform-server has its own library for parsing HTML, so we
sniff for that (by checking whether DOMParser exists) and fall back to
the standard strategy.

Thanks to @cure53 for the heads up on this issue.

mary-poppins commented:
You can preview 0bf4456 at https://pr17019-0bf4456.ngbuilds.io/.

mary-poppins commented:
You can preview 9ba9cea at https://pr17019-9ba9cea.ngbuilds.io/.

petebacondarwin commented:
Hurray! This is green. It needs a good security audit from someone like @IgorMinar or @mprobst (and maybe @cure53?) - especially the bits where we are generating HTML on the server.

mprobst commented:
/CC @rjamet

petebacondarwin commented on packages/platform-browser/test/security/html_sanitizer_spec.ts:
```diff
@@ -134,6 +135,32 @@ export function main() {
+                     // PlatformBrowser output
+                     '<p>&lt;img src=&#34;<img src="x"></p>' :
+                     // PlatformServer output
+                     '<p><img src="&lt;/style&gt;&lt;img src=x onerror=alert(1)//"></p>');
```
I believe that this HTML is safe to be sent to the browser but perhaps @cure53 can confirm?

rjamet replied:
Looks safe to me: garbage in a img src attribute shouldn't be an issue, and the HTML encoding seems right there too.

mhevery commented:
Is this ready to be merged? If so please add appropriate merge labels.

petebacondarwin commented:
@mhevery - No it needs further review - particularly because it is security sensitive. I am slightly worried about the HTML that might be produced by the platform-server.

mhevery commented:
OK, who is the next person who needs to review this?

petebacondarwin commented:
Perhaps you @mhevery ?? Otherwise I would say @mprobst or @IgorMinar.

petebacondarwin commented:
Or the people who are responsible for PlatformServer?

petebacondarwin commented:
I guess that this is no worse that what PlatformServer does right now.

vicb commented on packages/platform-browser/src/security/inert_body.ts:
```diff
@@ -0,0 +1,164 @@
+function isDOMParserAvailable() {
+  try {
+    return !!DOMParser;
+  } catch (e) {
```
Please refactor this code:

```
third_party/javascript/angular2/rc/packages/platform-browser/src/security/inert_body.closure.js:176: 
Originally at:
../../../third_party/javascript/angular2/rc/packages/platform-browser/src/security/inert_body.ts:161: ERROR - unreachable code
```

petebacondarwin replied:
Ok

petebacondarwin replied:
@vicb do you have any suggestions on how we could refactor this to make closure compiler happy?
It has been a while now but I believe that on some platforms accessing DOMParser actually throws an error.

mhevery replied:
```
/**
 * @suppress {uselessCode}
 */
function isDOMParserAvailable() {
  try {
    return !!DOMParser;
  } catch (e) {
    return false;
  }
} 
```

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%250A%2520*%2520%2540suppress%2520%257BuselessCode%257D%250A%2520*%252F%250Afunction%2520isDOMParserAvailable()%2520%257B%250A%2520%2520try%2520%257B%250A%2520%2520%2520%2520return%2520!!DOMParser%253B%250A%2520%2520%257D%2520catch%2520(e)%2520%257B%250A%2520%2520%2520%2520return%2520false%253B%250A%2520%2520%257D%250A%257D%2520%250AisDOMParserAvailable()%253B

petebacondarwin replied:
Unfortunately, the Google Closure compiler under ADVANCED optimisations then removes this function altogether, it seems. :-/

mhevery replied:
turns out that accessing it as `window.DOMParser` does the trick

```
/**
 * @suppress {uselessCode}
 */
function isDOMParserAvailable() {
  try {
    return !!window.DOMParser;
  } catch (e) {
    return false;
  }
}
console.log(isDOMParserAvailable());
```

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%250A%2520*%2520%2540suppress%2520%257BuselessCode%257D%250A%2520*%252F%250Afunction%2520isDOMParserAvailable()%2520%257B%250A%2520%2520try%2520%257B%250A%2520%2520%2520%2520return%2520!!window.DOMParser%253B%250A%2520%2520%257D%2520catch%2520(e)%2520%257B%250A%2520%2520%2520%2520return%2520false%253B%250A%2520%2520%257D%250A%257D%250Aconsole.log(isDOMParserAvailable())%253B

petebacondarwin replied:
Awesome, thanks

alxhub commented:
This is failing in Closure Compiler internally, with the following error:

1 error(s): packages/platform-browser/src/security/inert_body.closure.js:176: 
Originally at: platform-browser/src/security/inert_body.ts:161:

ERROR - unreachable code

danbrouwer commented:
Will this be merged to v5 as well?  This vulnerability is being flagged by McAfee Secure.

IgorMinar commented:
@danbrouwer we'll see in what shape this lands in master. if a breaking change is required then we won't be able to merge it into v5.

petebacondarwin commented:
> This is failing in Closure Compiler internally, with the following error:
> 
> 1 error(s): packages/platform-browser/src/security/inert_body.closure.js:176:
> Originally at: platform-browser/src/security/inert_body.ts:161:
> 
> ERROR - unreachable code

@alxhub or @IgorMinar - do you have suggestions how we can get round this, given that on some browsers accessing `window.DOMParser` actually does throw an error?

mary-poppins commented:
You can preview 1994ca1 at https://pr17019-1994ca1.ngbuilds.io/.

mary-poppins commented:
You can preview 74eab7f at https://pr17019-74eab7f.ngbuilds.io/.

mary-poppins commented:
You can preview 1188327 at https://pr17019-1188327.ngbuilds.io/.

mary-poppins commented:
You can preview 5f1556c at https://pr17019-5f1556c.ngbuilds.io/.

petebacondarwin commented on integration/_payload-limits.json:
```diff
@@ -3,7 +3,7 @@
       "uncompressed": {
         "inline": 1447,
-        "main": 151639,
+        "main": 154185,
```
Unfortunately, it appears these changes affect the size of the CLI bundle due to the increase in the `browser-platform` package.

petebacondarwin replied:
@mhevery / @IgorMinar who needs to OK this?

mhevery replied:
approved

IgorMinar replied:
OMG - 3kb?!?!? this makes me really sad. I suspect that we'll need to redo this in the future. lgtm for now.

mary-poppins commented:
You can preview 1cdc1a9 at https://pr17019-1cdc1a9.ngbuilds.io/.

mary-poppins commented:
You can preview 69dbd01 at https://pr17019-69dbd01.ngbuilds.io/.

mary-poppins commented:
You can preview eebc3c1 at https://pr17019-eebc3c1.ngbuilds.io/.

mhevery commented:
http://test/OCL:184883749:BASE:184883778:1518037784622:6aaeb0f0

IgorMinar reviewed (approved):
OMG - 3kb?!?!? this makes me really sad. I suspect that we'll need to redo this in the future. lgtm for now.

IgorMinar commented:
@petebacondarwin can you please send a PR against the LTS branch as well? I believe 4.4.x

petebacondarwin commented:
LTS version here: #22077
